models/uniswap/base/uniswap_v3_base_trades.sql

```diff
@@ -0,0 +1,109 @@
+{{ config(tags=['dunesql'],
+    schema = 'uniswap_v3_base',
+    alias = alias('trades'),
+    partition_by = ['block_month'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['block_date', 'blockchain', 'project', 'version', 'tx_hash', 'evt_index'],
+    post_hook='{{ expose_spells(\'["base"]\',
+                                "project",
+                                "uniswap_v3",
+                                \'["wuligy"]\') }}'
+    )
+}}
+
+{% set project_start_date = '2023-07-16' %}
+
+WITH dexs AS
+(
+    --Uniswap v3
+    SELECT
+        t.evt_block_time AS block_time
+        ,t.recipient AS taker
+        ,CAST(NULL as VARBINARY) as maker
+        ,CASE WHEN amount0 < INT256 '0' THEN abs(amount0) ELSE abs(amount1) END AS token_bought_amount_raw -- when amount0 is negative it means trader_a is buying token0 from the pool
+        ,CASE WHEN amount0 < INT256 '0' THEN abs(amount1) ELSE abs(amount0) END AS token_sold_amount_raw
+        ,NULL AS amount_usd
+        ,CASE WHEN amount0 < INT256 '0' THEN f.token0 ELSE f.token1 END AS token_bought_address
+        ,CASE WHEN amount0 < INT256 '0' THEN f.token1 ELSE f.token0 END AS token_sold_address
+        ,t.contract_address as project_contract_address
+        ,t.evt_tx_hash AS tx_hash
+        ,t.evt_index
+    FROM
+    {{ source('uniswap_v3_base', 'UniswapV3Pool_evt_Swap') }} t
+    INNER JOIN 
+    {{ source('uniswap_v3_base', 'UniswapV3Factory_evt_PoolCreated') }} f
+        ON f.pool = t.contract_address
+    {% if is_incremental() %}
+    WHERE t.evt_block_time >= date_trunc('day', now() - interval '7' day)
+    {% endif %}
+)
+SELECT
+    'base' AS blockchain
+    ,'uniswap' AS project
+    ,'3' AS version
+    ,TRY_CAST(date_trunc('DAY', dexs.block_time) AS date) AS block_date
+    ,CAST(date_trunc('month', dexs.block_time) AS date) AS block_month
+    ,dexs.block_time
+    ,erc20a.symbol AS token_bought_symbol
+    ,erc20b.symbol AS token_sold_symbol
+    ,case
+        when lower(erc20a.symbol) > lower(erc20b.symbol) then concat(erc20b.symbol, '-', erc20a.symbol)
+        else concat(erc20a.symbol, '-', erc20b.symbol)
+    end as token_pair
+    ,cast(dexs.token_bought_amount_raw as double) / power(10, erc20a.decimals) AS token_bought_amount
+    ,cast(dexs.token_sold_amount_raw as double) / power(10, erc20b.decimals) AS token_sold_amount
```

Comment:
i pulled up the uniswap v3 ethereum spell side-by-side, i think we can remove cast from here altogether, to keep consistency. when we finalize downstream spells, such as `dex.trades`, we will find out quickly if data types differ on these upstream spells and can fix universally then

Reply:
Yes, the cast can be removed here. I thought your suggestion the day before was to refer to the `editor` field types, so I explicitly cast it as a double type here. 

Of course, before doing this, I looked at the approaches of other people.

But it's alright, if there are any issues, please let me know promptly. I will adjust according to your specifications. 

This is my first pr, and after going through this phase, things will become smoother in the future : )

Reply:
that was my fault, apologies for that and no worries on your end!

i realized now that the explorer was probably showing spark types.